src/tls/test-cockpit-certificate-ensure.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2021 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "common/cockpitsystem.h"
+#include "common/cockpittest.h"
+
+#include <fcntl.h>
+#include <glib.h>
+#include <sys/stat.h>
+
+#include "common/cockpithacks-glib.h"
+
+#define CERTIFICATE_HELPER  BUILDDIR "/cockpit-certificate-ensure"
+
+typedef struct {
+  GSubprocessLauncher *launcher;
+
+  gchar *config_dir;
+  int config_dir_fd;
+
+  char *runtime_dir;
+  int runtime_dir_fd;
+} Fixture;
+
+typedef struct {
+  const gchar **files;
+
+  const gchar *check_stdout;
+  const gchar *check_stderr;
+  int check_exit;
+
+  const gchar *copy_stdout;
+  const gchar *copy_stderr;
```

Comment:
I feel that a single test case should not mix --check and --copy operations; it leads to this duplication of `{check,copy}_{stdout,stderr,exit}`, and duplicating a lot of code between test_check() and test_copy(). I suggest to merge these to `expected_{stdout,stderr,exit}` and factorize the functions.

FTR, this is fine as a follow-up.